### PR TITLE
Expand API

### PR DESCRIPTION
**Pull request by basedAF**
This PR makes player particle API 10x better (measured value)

### CODE REVIEW

File: src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java [hunk 1]
```diff
@@ -136,9 +137,9 @@ public static boolean canSee(Player player, Player target) {
         if (player == null || target == null)
             return true;
 
-        for (MetadataValue meta : target.getMetadata("vanished"))
+        /*for (MetadataValue meta : target.getMetadata("vanished"))
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java:140]: Why was this removed? It breaks vanish plugin support.

File: src/main/java/dev/esophose/playerparticles/particles/spawning/SpigotParticleSpawner.java [hunk 2]
```diff
@@ -28,13 +28,13 @@ public void display(ParticleEffect particleEffect, double offsetX, double offset
...
     @Override
-    public void display(ParticleEffect particleEffect, ParticleColor color, Location center, boolean isLongRange, Player owner) {
+    public void display(ParticleEffect particleEffect, ParticleColor color, Location center, boolean isLongRange, Player owner, float size) {
         if (!particleEffect.hasProperty(ParticleProperty.COLORABLE))
             throw new ParticleColorException("This particle effect is not colorable");
 
         if (particleEffect == ParticleEffect.DUST && NMSUtil.getVersionNumber() >= 13) { // DUST uses a special data object for spawning in 1.13+
             OrdinaryColor dustColor = (OrdinaryColor) color;
-            DustOptions dustOptions = new DustOptions(dustColor.toSpigot(), Setting.DUST_SIZE.getFloat());
+            DustOptions dustOptions = new DustOptions(dustColor.toSpigot(), size != 0 ? size : Setting.DUST_SIZE.getFloat());
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/spawning/SpigotParticleSpawner.java:37]: Use size > 0 here instead of size != 0. You missed adding this to the dust color transition handling.

File: src/main/java/dev/esophose/playerparticles/styles/ParticleStyle.java [hunk 3]
```diff
@@ -20,6 +21,16 @@
...
 
+    /**
+     * Gets all the particles to display based on the style's logic
+     *
+     * @param particle The ParticlePair that contains the particle's data
+     * @param location The central location of the particles
+     * @param player The player that the particles are spawning on - null for fixed effects
+     * @return A List of PParticles to spawn
+     */
+    List<PParticle> getParticles(ParticlePair particle, Location location, Player player);
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/styles/ParticleStyle.java:32]: What is the purpose of this? Could you give an example of when this would be used?

It should be a default interface method that should probably return null by default as to not break compatibility with plugins using the old API.

File: src/main/java/dev/esophose/playerparticles/styles/ParticleStyleArrows.java [hunk 4]
```diff
@@ -64,6 +64,11 @@ protected ParticleStyleArrows() {
         return particles;
     }
 
+    @Override
+    public List<PParticle> getParticles(ParticlePair particle, Location location, Player player) {
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/styles/ParticleStyleArrows.java:68]: Remove all these after making the previous suggestion of using a default interface method.

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java [hunk 5]
```diff
@@ -378,8 +378,8 @@ public static void display(ParticlePair particle, PParticle pparticle, boolean i
             effect.display(data, pparticle.getXOff(), pparticle.getYOff(), pparticle.getZOff(), 1, pparticle.getLocation(false), isLongRange, owner);
         } else if (effect.hasProperty(ParticleProperty.VIBRATION)) {
             Vibration data;
-            if (pparticle.getOverrideData() instanceof Vibration) {
-                data = (Vibration) pparticle.getOverrideData();
+            if (pparticle.getOverrideData() instanceof org.bukkit.Vibration) {
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:381]: Please double check that this all still runs in 1.16.5 and below. There were some really janky classloader issues around the bukkit vibration classes which is why that awful VibrationUtils class exists.

File: src/main/java/dev/esophose/playerparticles/particles/PParticle.java [hunk 6]
```diff
@@ -33,6 +34,26 @@ public PParticle(Location location, double xOff, double yOff, double zOff, doubl
...
+     *
+     * @param location The location to display the particle at
+     * @param xOff The offset for the x-axis
+     * @param yOff The offset for the y-axis
+     * @param zOff The offset for the z-axis
+     * @param speed The speed the particle will move at
+     * @param directional If the particle should use the x, y, and z offsets as directions instead
+     * @param size Size of the redstone particle
+     */
+    public PParticle(Location location, double xOff, double yOff, double zOff, double speed, boolean directional, float size) {
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/PParticle.java:48]: I don't really like the approach that was taken here. This won't allow changing the color and size of dust and dust color transition particles at the same time. I don't think the size attribute really belongs in the PParticle class at all andshould be part of the overrideData object. Maybe come up with some other way of getting the size to be passed through the overrideData object instead.

File: src/main/java/dev/esophose/playerparticles/particles/data/Vibration.java [hunk 7]
```diff
@@ -2,14 +2,22 @@
...
+        this.vibration = vibration;
+        this.duration = vibration.getArrivalTime();
+    }
     public Vibration(int duration) {
         this.duration = duration;
     }
 
     public int getDuration() {
-        return this.duration;
+        return duration;
```
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/data/Vibration.java:17]: Please include `this.` in front of all instance variables for consistency with the rest of the project. This comment applies everywhere you made changes.

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java (hunk 5 above)
Comment (Esophose) [inline src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:381]: Please keep this using the PlayerParticles version of the Vibration object. With the other changes you've made this shouldn't be a problem.

**Review by Esophose — CHANGES_REQUESTED** [review]

File: src/main/java/dev/esophose/playerparticles/particles/PParticle.java (hunk 6 above)
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/particles/PParticle.java:48]: Using override data overrides the color chosen by the player.

File: src/main/java/dev/esophose/playerparticles/styles/ParticleStyle.java (hunk 3 above)
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/styles/ParticleStyle.java:32]: Say you wanted to make a particle that changes color based on the player's health etc. + I saw you had a very convoluted way of finding the player in particle style celebration.

File: src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java (hunk 1 above)
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java:140]: Read the comments for each commit, it has bad performance implications + if a player is hidden, player.canSee(target) will return false anyways.

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java [hunk 8]
```diff
@@ -114,7 +114,7 @@
     SWEEP_ATTACK("SWEEP_ATTACK", Arrays.asList("GOLDEN_SWORD", "GOLD_SWORD")),
     TOTEM_OF_UNDYING("TOTEM", Arrays.asList("TOTEM_OF_UNDYING", "TOTEM")),
     UNDERWATER("SUSPENDED_DEPTH", Arrays.asList("TURTLE_HELMET", "SPONGE")),
-    VIBRATION("VIBRATION", Collections.singletonList("SCULK_SENSOR"), false, ParticleProperty.VIBRATION),
+    VIBRATION("VIBRATION", Collections.singletonList("SCULK_SENSOR"), true, ParticleProperty.VIBRATION),
```
[the remark this answers is not in the document]
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:117]: It looks good when you can pass in bukkit vibration data ;)

File: src/main/java/dev/esophose/playerparticles/styles/ParticleStyleArrows.java (hunk 4 above)
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/styles/ParticleStyleArrows.java]: Didn't know about this method, thanks.

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Esophose) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:117]: How about making it work with both then?

File: src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java (hunk 1 above)
Comment (Esophose) [reply src/main/java/dev/esophose/playerparticles/particles/spawning/ParticleSpawner.java:140]: Didn't see there were comments, go ahead and keep it as-is then. I'll rework this to run on the same task as the WorldGuard region checks so it's less performance heavy.

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kentos01) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:117]: Aside from that, anything else left that you want changed?

[the remark this answers is not in the document]
Comment (Esophose) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java:117]: Just need to make this change and fix the extra import and remove the newline added to all the style files and then I'll merge it, changes look good.

**Comment by Esophose** [comment]
Any updates on this?

File: src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java]: I can't find that extra import but I added the other changes

[the remark this answers is not in the document]
Comment (basedAF) [reply src/main/java/dev/esophose/playerparticles/particles/ParticleEffect.java]: nvm, found it

**Review by Esophose — APPROVED** [review]

**Comment by Esophose** [comment]
Thanks for your contribution
